### PR TITLE
Support for receiving fragmented IPv4 packets

### PR DESCRIPTION
**Pull request by podhrmic**
Implemented as an optional feature. Provided tests and expanded `server.rs` example. The code should be self-explanatory. Tested with UDP and ICMP packets. Replaces #185

### CODE REVIEW

File: src/iface/ethernet.rs [hunk 1]
```diff
@@ -667,14 +684,93 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+                    }
+                    fragment.reset();
+                    return Ok(Some(Ipv4Packet::new_checked(&fragment.rx_buffer[0..front])?));
+                }
+
+                // not the last fragment
+                return Ok(None);
+        	},
+        	None => {
+        		return Ok(None); // TODO: throw an exception?
```
Comment (whitequark) [inline src/iface/ethernet.rs]: Also log it, with a different message indicating that fragmentation is not supported.

File: src/iface/fragments.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
+use storage::Assembler;
+use time::Instant;
+use managed::ManagedSlice;
+
+/// Default timeout duration
+const FRAGMENTATION_TIMEOUT_MS: i64 = 500;
+
+#[allow(dead_code)]
```
Comment (whitequark) [inline src/iface/fragments.rs]: Better not include the module at all if built without the fragmentation feature.

File: src/iface/fragments.rs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+pub struct Packet<'a, T: 'a> {
+    assembler: Assembler,
+    pub rx_buffer: ManagedSlice<'a, T>,
+    id: u16,
```
Comment (whitequark) [inline src/iface/fragments.rs]: `ident`

File: src/iface/fragments.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    last_used: Instant,
+    header: bool,
```
Comment (whitequark) [inline src/iface/fragments.rs]: `has_header`

File: src/iface/fragments.rs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+use time::Instant;
+use managed::ManagedSlice;
+
+/// Default timeout duration
+const FRAGMENTATION_TIMEOUT_MS: i64 = 500;
+
+#[allow(dead_code)]
+pub struct Packet<'a, T: 'a> {
+    assembler: Assembler,
+    pub rx_buffer: ManagedSlice<'a, T>,
```
Comment (whitequark) [inline src/iface/fragments.rs]: Why is this `pub`?

File: src/iface/fragments.rs [hunk 6]
```diff
@@ -0,0 +1,170 @@
+use storage::Assembler;
+use time::Instant;
+use managed::ManagedSlice;
+
+/// Default timeout duration
+const FRAGMENTATION_TIMEOUT_MS: i64 = 500;
+
+#[allow(dead_code)]
+pub struct Packet<'a, T: 'a> {
```
Comment (whitequark) [inline src/iface/fragments.rs]: Why is it generic over `T`? We'll never want to defragment anything except `u8`. This just results in silly `T: Clone` bounds on functions.

File: src/iface/fragments.rs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        let range = offset+self.header_len..offset+payload_len+self.header_len;
+        self.rx_buffer[range].clone_from_slice(&data[header_len..]);
+        Ok(())
+    }
+
+    /// Return `true` if the packet has all fragments,
+    /// and can be reassmbled, and `false` otherwise
+    pub fn check_contig_range(&self) -> bool {
+        if self.total_len != 0 {
+            match self.assembler.peek_front() {
```
Comment (whitequark) [inline src/iface/fragments.rs]: `if let Some(front) = self.assembler.peek_front()`

File: src/iface/fragments.rs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            id: 0,
+            last_used: Instant::from_millis(0),
+            header: false,
+            header_len: 0,
+            total_len: 0,
+        }
+    }
+
+    /// Reset packet
+    pub fn reset(&mut self) {
```
Comment (whitequark) [inline src/iface/fragments.rs]: I think `Packet` would be much cleaner if instead of these empty values and `set_id` it was defined like...
```
struct Packet {
  rx_buffer: ManagedSlice<'a, T>,
  state: PacketState  
}

enum PacketState {
  Empty,
  Assembling {
    assembler: ...,
    has_header: ...,
    etc
  }
}
```
and then instead of methods like `set_id` you would have methods for transition between states, i.e. `start(ident: u16)`, `add_fragment(...)`, `extract_packet()`, and calling a method in the state that isn't correct for it would panic. Currently, this class of logic errors is not guarded against at all.

File: src/iface/fragments.rs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+		        }		
+    		},
+    		#[cfg(any(feature = "std", feature = "alloc"))]
+    		ManagedSlice::Owned(ref mut packets) => {
+    			packets.push(new_packet);
+    		}
+    	}
+    }
+
+    /// Check if the set contains a packet with given `id`
```
Comment (whitequark) [inline src/iface/fragments.rs]: This isn't enough. You need to at least match the ident field with the source IP, because otherwise you're introducing a vulnerability: anyone who knows that there is a flow of fragmented packets has an 1/65536 chance of arbitrarily altering their contents. This is a 100% viable attack vector and very similar attacks (utilizing the port number, but the concept is the same) have been demonstrated against DNS two decades ago.

Comment (whitequark) [reply src/iface/fragments.rs]: Also, you probably should use a ManagedMap here instead of hardcoding a linear search.

File: src/phy/mod.rs [hunk 10]
```diff
@@ -265,3 +265,8 @@ pub trait TxToken {
     fn consume<R, F>(self, timestamp: Instant, len: usize, f: F) -> Result<R>
         where F: FnOnce(&mut [u8]) -> Result<R>;
 }
+
+
+/// Ethernet header length to be added to the MTU, since the header is not counted into MTU
+/// 6 bytes src addr, 6 bytes dst addr, 2 bytes EthType, (optional) 4 bytes IEEE 802.1Q Header
+pub const ETHERNET_HAEDER_MAX_LEN: usize = 18;
```
Comment (whitequark) [inline src/phy/mod.rs]: This doesn't belong in this pull request. Please submit it separately and we'll have a discussion about it; in particular, you're counting 802.1Q, but not QinQ or QinQinQ.

**Comment by whitequark** [comment]
Please *please* fix indentation. It makes the code very hard to read.

File: examples/server.rs [hunk 11]
```diff
@@ -17,6 +17,8 @@ use smoltcp::socket::SocketSet;
 use smoltcp::socket::{UdpSocket, UdpSocketBuffer, UdpPacketMetadata};
 use smoltcp::socket::{TcpSocket, TcpSocketBuffer};
 use smoltcp::time::{Duration, Instant};
+#[cfg(feature = "fragmentation-ipv4")]
+use smoltcp::iface::{FragmentsSet, FragmentedPacket};
```
[the remark this answers is not in the document]
Comment (podhrmic) [reply examples/server.rs]: Done. On a side note, fragmentation is currently a part of the default features. If I remove it, it makes building the server_fragmented example really verbose (you have to enumerate all features). Is there some better way how to do that?

File: src/iface/ethernet.rs [hunk 12]
```diff
@@ -55,27 +56,29 @@ pub struct Interface<'b, 'c, DeviceT: for<'d> Device<'d>> {
...
 /// exclusively). However, it is still possible to call methods on its `inner` field.
-struct InterfaceInner<'b, 'c> {
+struct InterfaceInner<'b, 'c, 'e> {
     neighbor_cache:         NeighborCache<'b>,
     ethernet_addr:          EthernetAddress,
     ip_addrs:               ManagedSlice<'c, IpCidr>,
     #[cfg(feature = "proto-ipv4")]
     ipv4_gateway:           Option<Ipv4Address>,
     device_capabilities:    DeviceCapabilities,
+    fragments:				Option<FragmentsSet<'e>>,
```
[the remark this answers is not in the document]
Comment (podhrmic) [reply src/iface/ethernet.rs]: Fixed.

File: src/iface/fragments.rs (hunk 2 above)
Comment (podhrmic) [reply src/iface/fragments.rs]: But I need to include `FragmentSet` because interface now contains `fragments:              Option<FragmentSet<'e>>`, except without enabled fragmentation it compiles away as it is always None

File: src/iface/fragments.rs (hunk 5 above)
Comment (podhrmic) [reply src/iface/fragments.rs]: because `process_ipv4_fragment` returns `return Ok(Some(Ipv4Packet::new_checked(&fragment.rx_buffer[0..front])?));`

So having a getter function that returns a slice from a given range seemed unnecessary. But I am happy to add it.

File: src/iface/ethernet.rs [hunk 13]
```diff
@@ -667,14 +684,93 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+				    	fragments: &'r mut Option<FragmentsSet<'e>>
+				    ) ->
+                     Result<Option<Ipv4Packet<&'r [u8]>>> 
+    {
+        match *fragments {
+        	Some(ref mut fragments) => {
+                // get an existing fragment or attempt to get a new one
+				let fragment = match fragments.get_packet(ipv4_packet.ident(), timestamp) {
+				    Some(frag) => frag,
+				    None => return Ok(None), // TODO: throw an exception?
```
[the remark this answers is not in the document]
Comment (podhrmic) [reply src/iface/ethernet.rs]: I meant return an error. Fixed.

File: src/iface/ethernet.rs (hunk 1 above)
Comment (podhrmic) [reply src/iface/ethernet.rs]: Fixed

File: src/iface/fragments.rs (hunk 3 above)
Comment (podhrmic) [reply src/iface/fragments.rs]: should be fixed

File: src/iface/fragments.rs (hunk 4 above)
Comment (podhrmic) [reply src/iface/fragments.rs]: fixed

File: src/iface/fragments.rs (hunk 5 above)
Comment (podhrmic) [reply src/iface/fragments.rs]: Fixed.

File: src/phy/mod.rs (hunk 10 above)
Comment (podhrmic) [reply src/phy/mod.rs]: This is needed to get the fragmentation to work. In Linux MTU refers to the max length of the ethernet payload. But smoltcp counts MTU as max length of the ethernet frame. Hence without this added ethernet length, all fragmented packets are truncated. I am happy to fix is more meaningfully later, but without this fix the fragmentation reassembly doesn't work.

**Comment by podhrmic** [comment]
Work in progress - will ping you once I have all your comments implemented.

File: examples/server.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (whitequark) [reply examples/server.rs]: I think it's fine to have fragmentation in default features as it's required in IPv4. API consumers can always not initialize the FragmentSet (which will happen by default), so this would only impact code size, and anyone conscious about code size should already turn off unused features.

File: src/iface/ethernet.rs [hunk 14]
```diff
@@ -393,7 +406,8 @@ impl<'b, 'c, DeviceT> Interface<'b, 'c, DeviceT>
                 Some(tokens) => tokens,
             };
             rx_token.consume(timestamp, |frame| {
-                inner.process_ethernet(sockets, timestamp, &frame).map_err(|err| {
+            	let mut fragments = ::std::mem::replace(&mut inner.fragments, None);
```
[the remark this answers is not in the document]
Comment (podhrmic) [reply src/iface/ethernet.rs]: Currently requires `std`. See my comments in the updated version with proposed solutions.

File: src/iface/fragments.rs (hunk 8 above)
Comment (podhrmic) [reply src/iface/fragments.rs]: Done.

File: src/iface/fragments.rs (hunk 9 above)
Comment (podhrmic) [reply src/iface/fragments.rs]: Now I am checking the id and src & dst addresses. 
ManagedMap is not possible here because each packet needs an underlying storage. So I cannot simply add a new packet to the map, unless I explicitly provide a storage for that packet.

**Comment by podhrmic** [comment]
@whitequark please see the new commit. The main issue I found out is that currently, fragmentation requires `std` for the memory swap. The proposed solution is in `ethernet.rs` at line 410.

**Comment by whitequark** [comment]
> The main issue I found out is that currently, fragmentation requires std for the memory swap.

Huh? [core::mem::replace](https://doc.rust-lang.org/nightly/core/mem/fn.replace.html) exists. But I strongly dislike this solution. The code in EthernetInterface is already awful, let's not make it worse.

File: src/iface/fragments.rs (hunk 9 above)
Comment (whitequark) [reply src/iface/fragments.rs]: Of course it is possible, you use a map between ident/addresses and the *index* into the packet storage.

File: src/phy/mod.rs (hunk 10 above)
Comment (whitequark) [reply src/phy/mod.rs]: Can you extract it into a separate PR that we merge before this one?

**Comment by podhrmic** [comment]
> Huh? core::mem::replace exists. But I strongly dislike this solution. The code in EthernetInterface is already awful, let's not make it worse.

Nice! I missed that one. Updated and tested. I am curious if you find a better solution, as I said, I spent quite some time navigating between lifetimes and borrows on this one.

**Comment by podhrmic** [comment]
And one more question - in my `server_fragmented` example, I have the following rx buffer for UDP data:
```Rust
let udp_rx_buffer = UdpSocketBuffer::new(vec![UdpPacketMetadata::EMPTY], vec![0; 9000]);
```

I can receive 2 reassembled packets, each 4096 bytes long, but the third one triggers this error:
```
[1523987506.48s] (iface::ethernet): cannot process ingress packet: buffer space exhausted
```

Am I missing something? Shouldn't the buffer be cleared once the packet is processed?A similar thing happens to the regular server example.

**Comment by whitequark** [comment]
Might be related to https://github.com/m-labs/smoltcp/pull/187 cc @astro

**Comment by whitequark** [comment]
Can you please write a test case for this, on top of `master`? I would really appreciate that.

**Comment by podhrmic** [comment]
OK, will give it a try.

**Comment by podhrmic** [comment]
Another thing - I am not sure how to handle the CI build fails. In short, fragmentation currently works only for ipv4, but Ethernet Interface needs FragmentsSet even when it is not used. Hence the build fails when not using ipv4...:-/

**Comment by whitequark** [comment]
You could conditionally include the entire module.
```
#[cfg(feature = "proto-ipv4")]
mod fragments;
```

File: src/wire/ipv4.rs [hunk 15]
```diff
@@ -273,7 +273,7 @@ impl<T: AsRef<[u8]>> Packet<T> {
             Err(Error::Truncated)
         } else if self.header_len() as u16 > self.total_len() {
             Err(Error::Malformed)
-        } else if len < self.total_len() as usize {
+        } else if len < self.total_len() as usize && self.dont_frag() {
```
Comment (whitequark) [inline src/wire/ipv4.rs]: This needs an added test.

File: src/iface/fragments.rs [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+            },
+            PacketState::Assembling{..} => {
+                &mut self.rx_buffer[start..end]
+            }
+        }
+    }
+}
+
+pub struct Set<'a> {
+	packets: ManagedSlice<'a, Packet<'a>>,
```
Comment (whitequark) [inline src/iface/fragments.rs]: I would like this to use a ManagedMap as I have described previously.

File: src/wire/ipv4.rs (hunk 15 above)
Comment (podhrmic) [reply src/wire/ipv4.rs]: Can you elaborate? If the packet is not fragmented, the check goes on as before. If it is fragmented, I skip this check (and verify the length upon assembly).

File: src/iface/fragments.rs (hunk 16 above)
Comment (podhrmic) [reply src/iface/fragments.rs]: With ManagedMap and std it goes as follows:

1. new fragment arrives
2. fragment is not in the FragmentSet, create a new FragmentedPacket (its storage backed by a `vec!`) and insert it with the key to the FragmentSet

Now, with `no_std`, you cannot create a new FragmentedPacket with a `vec!` - you need an array, and you cannot allocate it from within the `process_ipv4_fragment` function because of lifetimes. So all FragmentedPackets have to be initialized prior polling (hence ManagedSlice).

Does it make sense?

**Comment by m-labs-homu** [comment]
:umbrella: The latest upstream changes (presumably 6c363762845e69dd316617f7ae0a083b4e534d5b) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by m-labs-homu** [comment]
:umbrella: The latest upstream changes (presumably 2363f97834b3d529c6344f1840ab001645e853d2) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by podhrmic** [comment]
Closing in favor of #236

File: src/phy/mod.rs (hunk 10 above)
Comment (podhrmic) [reply src/phy/mod.rs]: See #237
